qiskit/aqua/operators/gradients/circuit_gradients/lin_comb.py

```diff
@@ -0,0 +1,791 @@
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+from qiskit.aqua.aqua_globals import AquaError
+from qiskit.aqua.operators import OperatorBase, ListOp, ComposedOp
+from qiskit.aqua.operators.operator_globals import Z, I, One, Zero
+from qiskit.aqua.operators.primitive_ops.primitive_op import PrimitiveOp
+from qiskit.aqua.operators.state_fns import StateFn, CircuitStateFn, DictStateFn, VectorStateFn
+from qiskit.circuit import Gate, Instruction, Qubit
+from qiskit.circuit import (QuantumCircuit, QuantumRegister, ParameterVector,
+                            ParameterExpression)
+from qiskit.circuit.controlledgate import ControlledGate
+from qiskit.circuit.library import SGate, SdgGate, HGate
+from qiskit.circuit.library.standard_gates import (CXGate, CYGate, CZGate,
+                                                   IGate, RXGate, RXXGate,
+                                                   RYGate, RYYGate, RZGate,
+                                                   RZXGate, RZZGate, PhaseGate,
+                                                   UGate, ZGate)
+from qiskit.quantum_info import partial_trace
+
+from .circuit_gradient import CircuitGradient
+from ..derivative_base import DerivativeBase
+
+
+class LinComb(CircuitGradient):
+    """Compute the state gradient d⟨ψ(ω)|O(θ)|ψ(ω)〉/ dω respectively the gradients of the
+    sampling probabilities of the basis states of
+    a state |ψ(ω)〉w.r.t. ω.
+    This method employs a linear combination of unitaries,
+    see e.g. https://arxiv.org/pdf/1811.11184.pdf
+    """
+
+    def convert(self,
+                operator: OperatorBase,
+                params: Optional[Union[ParameterExpression, ParameterVector,
+                                       List[ParameterExpression],
+                                       Tuple[ParameterExpression, ParameterExpression],
+                                       List[Tuple[ParameterExpression, ParameterExpression]]]]
+                = None,
+                ) -> OperatorBase:
+        """ Convert the given operator into an operator object that represents the gradient w.r.t.
+            params
+
+        Args:
+            operator: The operator we are taking the gradient of: ⟨ψ(ω)|O(θ)|ψ(ω)〉
+            params: The parameters we are taking the gradient wrt: ω
+                    If a ParameterExpression, ParameterVector or List[ParameterExpression] is given,
+                    then the 1st order derivative of the operator is calculated.
+                    If a Tuple[ParameterExpression, ParameterExpression] or
+                    List[Tuple[ParameterExpression, ParameterExpression]]
+                    is given, then the 2nd order derivative of the operator is calculated.
+
+        Returns:
+            An operator corresponding to the gradient resp. Hessian. The order is in accordance with
+            the order of the given parameters.
+        """
+
+        return self._prepare_operator(operator, params)
```

Comment:
Is there a reason this is delegated to this single method? If no other logic is involved we probably don't need to additionally encapsulate this

Reply:
The reason is that we use this function to traverse through the given operator and thereby adapt the elements of the `operator.oplist` according to their type. So (Circuit)StateFns are adapted according to the insertion principles and measurements are extended by an additional measurement on the working qubit.